### PR TITLE
[LWDM] feat(coin-tron): expose energy/bandwidth breakdown and make TRC20 fee energy-aware

### PR DESCRIPTION
**Pull request by lysyi3m**
### 📝 Description

TRON's send flow shows a flat TRC20 fee (`STANDARD_FEES_TRC_20`, ~13.74 TRX) regardless of the sender's staked energy, and exposes no resource data — so the UI cannot tell the user whether a transfer is covered by energy/bandwidth or paid by burning TRX.

This exposes four `BigNumber` fields on the Tron `TransactionStatus` (`energyRequired`, `energyAvailable`, `bandwidthRequired`, `bandwidthAvailable`) and makes the active-recipient TRC20 fee energy-aware: 0 TRX when energy and bandwidth cover the transfer, the real shortfall cost otherwise (derived from the `triggerConstantContract` simulation via the existing `computeEnergyFee` / `computeBandwidthFee` helpers), falling back to the previous flat fee on any uncertainty. `energyRequired` uses the full simulated `energy_used` — the caller pays 100% in practice (`origin_energy_usage = 0` on mainnet, incl. USDT), so the `consume_user_resource_percent` sponsorship ratio is not applied. The `TronNotEnoughEnergy` warning now uses the real `energyRequired > energyAvailable` comparison instead of the hardcoded `47619` heuristic.

The insufficient-gas check now raises `NotEnoughGas` only when a fee is actually owed — a 0-TRX account whose transfer is fully covered by staked resources is no longer blocked, while the buy-currency deeplink still fires whenever a fee > 0. Native TRX / TRC10 fee amounts and the new-recipient paths are otherwise unchanged. Prerequisite for the LWD/LWM send-flow fee tooltip & drawer.

### 🔗 Context

- **JIRA / GitHub issue**: [LIVE-34149](https://ledgerhq.atlassian.net/browse/LIVE-34149)
- **ADR** (if any): —

[LIVE-34149]: https://ledgerhq.atlassian.net/browse/LIVE-34149?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts [hunk 1]
```diff
@@ -25,11 +28,115 @@ const getFeesFromBandwidth = (account: Account, transaction: Transaction): BigNu
...
+        energyLimit: zero,
+      },
+    };
+  }
+
+  const { freeUsed, freeLimit, gainedUsed, gainedLimit } = extractBandwidthInfo(networkInfo);
+  const bandwidthAvailable = freeLimit.minus(freeUsed).plus(gainedLimit).minus(gainedUsed);
+  const bandwidthRequired = getEstimatedBlockSize(account, transaction);
+  const energyAvailable = networkInfo.energyLimit.minus(networkInfo.energyUsed);
+
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts:98]: Fixed — `bandwidthAvailable` and `energyAvailable` are now clamped with `BigNumber.maximum(0, …)`, with a regression test covering the used > limit case.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts [hunk 2]
```diff
@@ -82,10 +232,11 @@ const getEstimatedFees = async (
   account: Account,
   transaction: Transaction,
   tokenAccount?: TokenAccount | null,
+  breakdown?: FeeResourceBreakdown,
 ) => {
   const feesFromAccountActivation =
     transaction.mode === "send"
-      ? await getFeesFromAccountActivation(account, transaction, tokenAccount)
+      ? await getFeesFromAccountActivation(account, transaction, tokenAccount, breakdown)
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts]: Fixed — `getEstimatedFees` now routes the activation/bandwidth paths through `breakdown.networkInfo` when a breakdown is supplied (`txForFees`), so their gating uses the same network data as the breakdown.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts [hunk 3]
```diff
@@ -70,9 +182,47 @@ const getFeesFromAccountActivation = async (
     }
   }
 
-  // if account is activated and it does already have a trc20 balance for given token.
+  // Active-recipient TRC20: energy-aware fee (was an unconditional flat constant). Any uncertainty
+  // falls back to STANDARD_FEES_TRC_20. Inactive recipients keep the flat constant unchanged — the
+  // energy-aware path only covers the active-recipient case it replaced.
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts]: Fixed — reworded the comment to state accurately that inactive recipients with insufficient bandwidth are handled by the new-recipient branch above, and the sufficient-bandwidth case keeps the flat constant.

File: libs/coin-modules/coin-tron/src/logic/estimateFees.ts [hunk 4]
```diff
@@ -70,7 +70,7 @@ const computeBandwidthFee = (
   return missing.multipliedBy(params.transactionFee);
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/logic/estimateFees.ts]: Fixed — `computeBandwidthFee` now clamps available to ≥0 before computing the shortfall, matching the breakdown. Added a regression test for the used > limit case.

File: libs/coin-modules/coin-tron/src/logic/estimateFees.ts [hunk 5]
```diff
@@ -70,7 +70,7 @@ const computeBandwidthFee = (
...
 };
 
-const computeEnergyFee = (
+export const computeEnergyFee = (
   energyNeeded: number,
   networkInfo: NetworkInfo,
   params: ChainParameters,
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/logic/estimateFees.ts:90]: Fixed — `computeEnergyFee` now clamps available energy to ≥0 before the shortfall, consistent with the breakdown. Added a regression test for energyUsed > energyLimit.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts [hunk 6]
```diff
@@ -25,11 +28,120 @@ const getFeesFromBandwidth = (account: Account, transaction: Transaction): BigNu
...
+
+  try {
+    const transactionIntent: TransactionIntent = {
+      intentType: "transaction",
+      type: transaction.mode,
+      sender: account.freshAddress,
+      recipient: transaction.recipient,
+      amount: BigInt(transaction.amount.toString()),
+      asset: { type: "trc20", assetReference: tokenAccount.token.contractAddress },
+    };
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts:124]: Fixed — the TRC20 simulation now uses `.toFixed()` (no exponential notation, so `BigInt()` handles large amounts), and for `useAllAmount` it simulates with the token balance instead of a 0-value transfer. Both covered by new tests.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts [hunk 7]
```diff
@@ -25,11 +28,124 @@ const getFeesFromBandwidth = (account: Account, transaction: Transaction): BigNu
...
+    freeLimit.minus(freeUsed).plus(gainedLimit).minus(gainedUsed),
+  );
+  const bandwidthRequired = getEstimatedBlockSize(account, transaction);
+  const energyAvailable = BigNumber.maximum(0, networkInfo.energyLimit.minus(networkInfo.energyUsed));
+
+  // No energy dimension (native/TRC10), or a zero-amount send that is not "send max" — the latter is
+  // headed for an AmountRequired error, so skip the on-chain simulation instead of spending a call
+  // on it (getTransactionStatus computes the breakdown before amount validation runs).
+  const isZeroAmountSend = transaction.amount.isZero() && !transaction.useAllAmount;
+  if (!tokenAccount || tokenAccount.token.tokenType !== "trc20" || isZeroAmountSend) {
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts]: Fixed — `getFeeResourceBreakdown` now runs the energy simulation only for `mode === "send"`, so non-send transactions with a token sub-account no longer trigger an irrelevant transfer simulation. Added a test with a freeze transaction.

File: libs/coin-modules/coin-tron/src/logic/estimateFees.ts [hunk 8]
```diff
@@ -58,24 +58,25 @@ const estimateEnergy = async (intent: TransactionIntent<TronMemo>): Promise<numb
...
   size: number,
   networkInfo: NetworkInfo,
   params: ChainParameters,
 ): BigNumber => {
   const freeAvailable = networkInfo.freeNetLimit.minus(networkInfo.freeNetUsed);
   const stakedAvailable = networkInfo.netLimit.minus(networkInfo.netUsed);
-  const available = freeAvailable.plus(stakedAvailable);
+  // Clamp to ≥ 0: a node reporting used > limit must not inflate the shortfall beyond `size`.
+  const available = BigNumber.maximum(0, freeAvailable.plus(stakedAvailable));
   const missing = BigNumber.maximum(0, new BigNumber(size).minus(available));
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/logic/estimateFees.ts]: Fixed — `computeBandwidthFee` now clamps each bucket (`max(0, free) + max(0, staked)`) before summing, so a negative free bucket no longer reduces staked. Added a test with a negative-free/positive-staked case.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts [hunk 9]
```diff
@@ -25,11 +28,130 @@ const getFeesFromBandwidth = (account: Account, transaction: Transaction): BigNu
...
+    };
+  }
+
+  const { freeUsed, freeLimit, gainedUsed, gainedLimit } = extractBandwidthInfo(networkInfo);
+  // Clamp availability to ≥ 0: the node can report used > limit, and a negative "available" is
+  // nonsensical for the UI and would over-state the shortfall.
+  const bandwidthAvailable = BigNumber.maximum(
+    0,
+    freeLimit.minus(freeUsed).plus(gainedLimit).minus(gainedUsed),
+  );
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts:92]: Fixed — `bandwidthAvailable` in the breakdown now clamps each bucket independently too, matching `computeBandwidthFee`. Covered by a test.

File: libs/coin-modules/coin-tron/src/types/bridge.ts [hunk 10]
```diff
@@ -298,5 +298,13 @@ export function isTronAccount(account: Account): account is TronAccount {
...
+// Additive resource-breakdown fields for the send-flow UI: sufficiency is `energyRequired <=
+// energyAvailable && bandwidthRequired <= bandwidthAvailable`. On a failed energy estimate,
+// energyRequired is a safe insufficient sentinel. Not mirrored to TransactionStatusRaw.
+export type TransactionStatus = TransactionStatusCommon & {
+  energyRequired: BigNumber;
+  energyAvailable: BigNumber;
+  bandwidthRequired: BigNumber;
+  bandwidthAvailable: BigNumber;
+};
 export type TransactionStatusRaw = TransactionStatusCommonRaw;
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/types/bridge.ts:310]: Fixed — added a custom `fromTransactionStatusRaw` that defaults the four resource fields to 0 on deserialization, so the required `BigNumber` fields are never `undefined` (they are live-derived and intentionally not persisted). Follows the same custom-serializer approach coin-bitcoin uses for its extra status fields. Covered by a new transaction.test.ts.

File: libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.ts [hunk 11]
```diff
@@ -189,10 +188,24 @@ const getTransactionStatus = async (
     }
   }
 
-  const estimatedFees =
-    Object.entries(errors).length > 0
-      ? new BigNumber(0)
-      : await getEstimatedFees(acc, transaction, tokenAccount);
+  // Skip the breakdown (and its network calls) for invalid transactions; reuse the single result
+  // for both the status fields and the fee so they can't disagree.
+  const hasErrors = Object.entries(errors).length > 0;
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.ts]: Fixed — reworded to state accurately that the breakdown runs when no error has been detected so far (the recipient/resource checks above), noting that amount validation runs later and the zero-amount case is guarded inside getFeeResourceBreakdown.

File: libs/coin-modules/coin-tron/src/types/bridge.ts [hunk 12]
```diff
@@ -298,5 +298,13 @@ export function isTronAccount(account: Account): account is TronAccount {
 }
 export type TronAccount = Account & { tronResources: TronResources };
 export type TronAccountRaw = AccountRaw & { tronResources: TronResourcesRaw };
-export type TransactionStatus = TransactionStatusCommon;
+// Additive resource-breakdown fields for the send-flow UI: sufficiency is `energyRequired <=
+// energyAvailable && bandwidthRequired <= bandwidthAvailable`. On a failed energy estimate,
+// energyRequired is a safe insufficient sentinel. Not mirrored to TransactionStatusRaw.
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/types/bridge.ts]: Fixed — reworded the comment so the sufficiency expression is no longer wrapped in a split backtick span.

File: libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.ts [hunk 13]
```diff
@@ -189,10 +188,25 @@ const getTransactionStatus = async (
...
+    energyRequired = resourceBreakdown.energyRequired;
+    energyAvailable = resourceBreakdown.energyAvailable;
+    bandwidthRequired = resourceBreakdown.bandwidthRequired;
+    bandwidthAvailable = resourceBreakdown.bandwidthAvailable;
+    estimatedFees = await getEstimatedFees(acc, transaction, tokenAccount, resourceBreakdown);
+  }
+
   const balance =
     account.type === "Account"
       ? BigNumber.max(0, account.spendableBalance.minus(estimatedFees))
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.ts:212]: Good catch — confirmed a real typo (`undelegateResource` vs the canonical `unDelegateResource`). It is pre-existing (introduced 2024-06-21) on a line this PR does not modify, and fixing it changes `unDelegateResource` totalSpent behaviour unrelated to this energy/fee change, so per scope discipline it is tracked separately in [LIVE-34199](https://ledgerhq.atlassian.net/browse/LIVE-34199) rather than folded in here.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts [hunk 14]
```diff
@@ -1,10 +1,13 @@
...
 import BigNumber from "bignumber.js";
 import { estimateFees, getAccount } from "../logic";
 import { ACTIVATION_FEES, STANDARD_FEES_NATIVE, STANDARD_FEES_TRC_20 } from "../logic/constants";
-import type { AccountTronAPI } from "../network/types";
-import type { Transaction } from "../types";
+import { computeBandwidthFee, computeEnergyFee, estimateEnergy } from "../logic/estimateFees";
+import { getChainParameters, getTronAccountNetwork } from "../network";
+import type { AccountTronAPI, ChainParameters } from "../network/types";
+import type { NetworkInfo, Transaction } from "../types";
 import { extractBandwidthInfo, getEstimatedBlockSize } from "./utils";
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts:12]: Good catch — clamped each bandwidth bucket (free and staked/gained) to ≥0 before summing in `getFeesFromBandwidth`, mirroring the resource-breakdown computation so the exposed `bandwidthAvailable` and the charged fee cannot disagree. Added a regression test for the negative-free-bucket case.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts [hunk 15]
```diff
@@ -70,9 +232,13 @@ const getFeesFromAccountActivation = async (
     }
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts:241]: This is pre-existing code, unchanged by this PR, in the inactive-recipient activation path (itself guarded upstream by `TronSendTrc20ToNewAccountForbidden`). On the exponential-notation and send-max concerns it fails safe: `estimateFees` catches and falls back to the flat `STANDARD_FEES_TRC_20`, so it over-reports rather than under-reporting. Keeping this PR scoped to the energy/bandwidth breakdown and tracking the hardening (toFixed + deriving the simulated amount from the token balance when `useAllAmount`, matching `getFeeResourceBreakdown`) as a separate follow-up.

File: libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.test.ts [hunk 16]
```diff
@@ -0,0 +1,390 @@
+import {
+  AmountRequired,
+  InvalidAddress,
+  InvalidAddressBecauseDestinationIsAlsoSource,
+  NotEnoughBalance,
+  NotEnoughGas,
+} from "@ledgerhq/errors";
+import { TokenAccount } from "@ledgerhq/types-live";
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.test.ts]: Done — switched to `import type` for this type-only import.

File: libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.test.ts [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+  fetchTronAccount,
+  fetchTronContract,
+  getChainParameters,
+  getDelegatedResource,
+  getTronAccountNetwork,
+  getTronSuperRepresentatives,
+  triggerConstantContract,
+} from "../network";
+import { STANDARD_FEES_TRC_20 } from "../logic/constants";
+import { NetworkInfo, Transaction, TronAccount } from "../types";
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.test.ts]: Done — switched to `import type` for these type-only imports.

File: libs/coin-modules/coin-tron/src/test/bridgeDatasetTest.ts [hunk 18]
```diff
@@ -38,6 +39,27 @@ const getTokenAccountId = (account: Account, id: string) =>
...
+  const bandwidthShortfall = BigNumber.maximum(
+    0,
+    status.bandwidthRequired.minus(status.bandwidthAvailable),
+  );
+  const energyShortfall = BigNumber.maximum(0, status.energyRequired.minus(status.energyAvailable));
+  const expectedFees = bandwidthShortfall
+    .multipliedBy(params.transactionFee)
+    .plus(energyShortfall.multipliedBy(params.energyFee));
+  expect(status.estimatedFees.toFixed()).toEqual(expectedFees.toFixed());
+};
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/test/bridgeDatasetTest.ts]: Removed the `any`. Rather than annotate `expect` (the harness types the callback's `expect` arg as a loose function that is not assignable to the full `jest.Expect` interface), I refactored the helper into a pure `expectedEnergyAwareTrc20Fee(status): Promise<BigNumber>` that returns the expected fee, and moved the assertion to the call sites where `expect` is already contextually typed by the dataset harness. No `any` remains.

File: libs/coin-modules/coin-tron/src/test/bridgeDatasetTest.ts [hunk 19]
```diff
@@ -38,6 +39,24 @@ const getTokenAccountId = (account: Account, id: string) =>
...
+  const params = await getChainParameters();
+  const bandwidthShortfall = BigNumber.maximum(
+    0,
+    status.bandwidthRequired.minus(status.bandwidthAvailable),
+  );
+  const energyShortfall = BigNumber.maximum(0, status.energyRequired.minus(status.energyAvailable));
+  return bandwidthShortfall
+    .multipliedBy(params.transactionFee)
+    .plus(energyShortfall.multipliedBy(params.energyFee));
+};
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/test/bridgeDatasetTest.ts]: Good call — added an early return: when both the bandwidth and energy shortfalls are zero the fee is 0 regardless, so `getChainParameters()` is now only fetched when there is an actual shortfall to price.

File: libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.test.ts [hunk 20]
```diff
@@ -0,0 +1,390 @@
...
+        createTransaction({ amount: new BigNumber(0) }),
+      );
+      expect(status.errors.amount).toBeInstanceOf(AmountRequired);
+    });
+
+    it("send exceeding the spendable balance → NotEnoughBalance", async () => {
+      const status = await getTransactionStatus(
+        createAccount({ balance: new BigNumber(50_000_000), spendableBalance: new BigNumber(50_000_000) }),
+        createTransaction({ amount: new BigNumber(100_000_000) }),
+      );
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.test.ts:309]: Ran the package's formatter (`oxfmt`, the repo's configured tool for coin-tron) over the file, so the long lines are now wrapped to the repo style.

File: libs/coin-modules/coin-tron/src/test/bridgeDatasetTest.ts [hunk 21]
```diff
@@ -38,6 +39,34 @@ const getTokenAccountId = (account: Account, id: string) =>
...
+  // No shortfall means no fee, so skip the chain-parameters fetch entirely.
+  let expectedFees = new BigNumber(0);
+  if (!(bandwidthShortfall.isZero() && energyShortfall.isZero())) {
+    const params = await getChainParameters();
+    expectedFees = bandwidthShortfall
+      .multipliedBy(params.transactionFee)
+      .plus(energyShortfall.multipliedBy(params.energyFee));
+  }
+  expect(status.estimatedFees.toFixed()).toEqual(expectedFees.toFixed());
+};
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/test/bridgeDatasetTest.ts:78]: Great catch — you're right, the assertion was wrong for the simulation-failure path (production returns the flat `STANDARD_FEES_TRC_20` while the breakdown reports the `energyRequired = energyAvailable + 1` sentinel), which would have made this test flap on transient node errors. Fixed: the helper now detects that sentinel (an exact +1 gap is unreachable for a real transfer, whose energy need is in the thousands) and asserts the flat fee there, keeping exact shortfall equality only for a genuine estimate.

File: libs/coin-modules/coin-tron/src/test/bridgeDatasetTest.ts [hunk 22]
```diff
@@ -38,6 +39,43 @@ const getTokenAccountId = (account: Account, id: string) =>
...
+  expect: ExpectToEqual,
+  statusCommon: unknown,
+): Promise<void> => {
+  const status = statusCommon as TransactionStatus;
+  // When the live energy simulation fails, production falls back to the flat STANDARD_FEES_TRC_20
+  // and the breakdown reports an "insufficient" sentinel (energyRequired = energyAvailable + 1). A
+  // real transfer needs thousands of energy units, so an exact +1 gap only ever means the sentinel:
+  // assert the pessimistic flat fee there rather than pricing it as a 1-unit shortfall, which would
+  // otherwise make this test flap on transient node errors.
+  if (status.energyRequired.eq(status.energyAvailable.plus(1))) {
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/test/bridgeDatasetTest.ts]: You're right — the `+1` gap is not a reliable discriminator (a genuine 1-unit shortfall lands on it too), so the sentinel check could misprice a boundary case and flap. Reworked it to discriminate on the fee itself: if `estimatedFees` equals the flat `STANDARD_FEES_TRC_20` it is treated as the uncertain-estimate fallback and accepted; otherwise it must equal the shortfall priced from the exposed breakdown. Any third value (e.g. a silent regression to a too-low fee) still fails. The fragile `energyRequired == energyAvailable + 1` heuristic is gone.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts [hunk 23]
```diff
@@ -25,11 +33,170 @@ const getFeesFromBandwidth = (account: Account, transaction: Transaction): BigNu
...
+      err,
+    });
+    return {
+      energyRequired: energyAvailable.plus(1),
+      energyAvailable,
+      bandwidthRequired,
+      bandwidthAvailable,
+      energyEstimated: false,
+      networkInfo,
+    };
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts:150]: This is by design and fails safe. The `energyEstimated` flag is deliberately kept internal to the coin module (it was removed from the public `TransactionStatus` to avoid a messy fifth field). On a failed estimate the breakdown reports the insufficient sentinel AND `estimatedFees` falls back to the pessimistic flat `STANDARD_FEES_TRC_20`, so the direction is always over-report + a `TronNotEnoughEnergy` warning — never an under-report. The only consequence is a rare, transient-failure UI edge (a ~1-unit shortfall shown alongside the flat fee); re-exposing the flag would reverse the agreed design for no safety gain. Leaving as-is intentionally.

File: libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.ts [hunk 24]
```diff
@@ -219,25 +233,14 @@ const getTransactionStatus = async (
...
     ) {
-      const contractUserEnergyConsumption = await getContractUserEnergyRatioConsumption(
-        account.token.contractAddress,
-      );
-
-      if (contractUserEnergyConsumption > 0) {
-        warnings.amount = new TronNotEnoughEnergy();
-      }
+      warnings.amount = new TronNotEnoughEnergy();
     }
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getTransactionStatus.ts:244]: Good catch — real regression the energy-aware path introduced. Fixed: the gas check now only requires TRX when a fee is actually owed — `parentAccountBalance.lt(estimatedFees) || (parentAccountBalance.isZero() && estimatedFees.gt(0))`. A 0-TRX account whose transfer is fully covered by staked resources (fee 0) is no longer blocked. Added a regression test; the buy-currency deeplink is preserved whenever a fee > 0 is owed.

File: libs/coin-modules/coin-tron/src/logic/estimateFees.ts [hunk 25]
```diff
@@ -16,7 +16,7 @@ import { ACTIVATION_FEES, STANDARD_FEES_NATIVE, STANDARD_FEES_TRC_20 } from "./c
 // Byte sizes of the fully signed transaction (raw_data + signature + protobuf wrapping).
 // Mirrors bridge/utils.ts:getEstimatedBlockSize. Using craftTransaction's raw_data_hex alone
 // would underestimate by ~50% because it doesn't include the signature/outer envelope.
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/logic/estimateFees.ts]: Clarified the comment: it now notes that `estimatedTxSize` only approximates `getEstimatedBlockSize`, and that the `vote` size is a flat estimate because `TransactionIntent` carries no vote count — so it does not grow per vote (a minor bandwidth under-estimate for multi-vote intents). The classic bridge vote path still uses `getEstimatedBlockSize`, so the fee shown to users is unaffected.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts [hunk 26]
```diff
@@ -25,11 +33,170 @@ const getFeesFromBandwidth = (account: Account, transaction: Transaction): BigNu
...
+      networkInfo: {
+        family: "tron",
+        freeNetUsed: zero,
+        freeNetLimit: zero,
+        netUsed: zero,
+        netLimit: zero,
+        energyUsed: zero,
+        energyLimit: zero,
+      },
+    };
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts]: Fixed. The network-info fetch-failure path now computes `needsEnergySim` (TRC20 + send + non-zero-amount) up front and only emits the `energyRequired = energyAvailable + 1` sentinel when energy is genuinely required. Native/TRC10 (and non-transfer modes) now report `energyRequired: 0` and `energyEstimated: true` even when the fetch fails, so the exposed breakdown no longer shows a phantom energy shortfall for zero-energy currencies. Added a regression test for the native fetch-failure case.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.test.ts [hunk 27]
```diff
@@ -2,14 +2,22 @@ import { Account, TokenAccount } from "@ledgerhq/types-live";
 import BigNumber from "bignumber.js";
 import { estimateFees, getAccount } from "../logic";
 import { ACTIVATION_FEES, STANDARD_FEES_NATIVE, STANDARD_FEES_TRC_20 } from "../logic/constants";
+import { computeBandwidthFee, computeEnergyFee, estimateEnergy } from "../logic/estimateFees";
+import { getChainParameters, getTronAccountNetwork } from "../network";
+import type { ChainParameters } from "../network/types";
 import { Transaction } from "../types";
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.test.ts]: Done — changed to `import type { Transaction }`.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.test.ts [hunk 28]
```diff
@@ -2,14 +2,22 @@ import { Account, TokenAccount } from "@ledgerhq/types-live";
 import BigNumber from "bignumber.js";
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.test.ts]: Done — switched to `import type { Account, TokenAccount }`; both are only used in type positions here.

File: libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts [hunk 29]
```diff
@@ -1,10 +1,13 @@
 import type { TransactionIntent } from "@ledgerhq/coin-module-framework/api/index";
+import { log } from "@ledgerhq/logs";
 import { Account, TokenAccount } from "@ledgerhq/types-live";
 import BigNumber from "bignumber.js";
```
[the remark this answers is not in the document]
Comment (lysyi3m) [reply libs/coin-modules/coin-tron/src/bridge/getEstimateFees.ts:4]: Done — switched to `import type { Account, TokenAccount }`; both are type-only in this module.

**Review by francois-guerin-ledger — APPROVED** [review]
